include/unifex/variant_sender.hpp

```diff
@@ -0,0 +1,124 @@
+ * limitations under the License.
+ */
+#pragma once
+
+#include <unifex/config.hpp>
+#include <unifex/receiver_concepts.hpp>
+#include <unifex/sender_concepts.hpp>
+#include <unifex/blocking.hpp>
+#include <unifex/std_concepts.hpp>
+
+#include <exception>
+#include <tuple>
+#include <type_traits>
+#include <utility>
+#include <variant>
+
+#include <unifex/detail/prologue.hpp>
+
+namespace unifex {
+namespace _variant_sender {
+
+template <typename... Ops>
+struct _op {
+  struct type;
+};
+template <typename... Ops>
+using operation = typename _op<Ops...>::type;
+
+template <typename... Ops>
+struct _op<Ops...>::type {
+  void start() & noexcept {
+    std::visit([](auto& op) noexcept(noexcept(unifex::start(op))) { unifex::start(op); }, variant_op);
+  }
+
+  std::variant<Ops...> variant_op;
+};
+
+
+template <_block::_enum First, _block::_enum... Rest>
+struct max_blocking_kind {
+  constexpr auto operator()() noexcept { return First; }
+};
+
+template <_block::_enum First, _block::_enum Second, _block::_enum... Rest>
+struct max_blocking_kind<First, Second, Rest...> {
+  constexpr auto operator()() noexcept {
+    if constexpr (First == Second) {
+      return max_blocking_kind<First, Rest...>{}();
+    } else if constexpr (
+        (First == blocking_kind::always &&
+        Second == blocking_kind::always_inline) ||
+        (Second == blocking_kind::always &&
+        First == blocking_kind::always_inline)) {
+      return max_blocking_kind<blocking_kind::always_inline, Rest...>{}();
+    } else {
+      return blocking_kind::maybe;
+    }
+  }
+};
+
+template <typename... Senders>
+struct _sender {
+  class type;
+};
+template <typename... Senders>
+using sender = typename _sender<Senders...>::type;
+
+template <typename... Senders>
+class _sender<Senders...>::type {
+  std::variant<Senders...> sender_variant;
+
+ public:
+
+  template <
+      template <typename...> class Variant,
+      template <typename...> class Tuple>
+  using value_types = typename concat_type_lists_unique_t<sender_value_types_t<Senders, type_list, Tuple>...>::template apply<Variant>;
+
+  template <template <typename...> class Variant>
+  using error_types = concat_type_lists_unique_t<sender_error_types_t<Senders, Variant>...>;
+
+  static constexpr bool sends_done = std::disjunction_v<std::bool_constant<sender_traits<Senders>::sends_done>...>;
+
+  template<typename ConcreteSender>
+  type(ConcreteSender&& concrete_sender)
+    noexcept(std::is_nothrow_constructible_v<std::variant<Senders...>, decltype(concrete_sender)>)
+    : sender_variant(std::forward<ConcreteSender>(concrete_sender)) {}
+
+  template(typename This, typename Receiver)
+    (requires same_as<remove_cvref_t<This>, type> AND std::conjunction_v<std::bool_constant<sender_to<member_t<This, Senders>, Receiver>>...>)
+  friend auto tag_invoke(tag_t<connect>, This&& that, Receiver&& r)
+    noexcept(std::conjunction_v<unifex::is_nothrow_connectable<Senders, Receiver>...>)
```

Comment:
I think this `noexcept` clause is imprecise.  I think the member variant ought to be forwarded into `visit`, which means the `connect` expression may be given either an lvalue or an rvalue for any of the constituent senders.

I wonder what the value categories end up being in the cross product of (`This` is an value, `This` is an rvalue) x (one of `Senders` is a plain type, one of `Senders` is an lvalue reference, one of `Senders` is an rvalue reference).

Reply:
Seems that [`std::variant` isn't permitted to hold references](https://en.cppreference.com/w/cpp/utility/variant):
> A variant is not permitted to hold references, arrays, or the type void. Empty variants are also ill-formed (std::variant<[std::monostate](http://en.cppreference.com/w/cpp/utility/variant/monostate)> can be used instead).

So I'll add some tests for when `This` is an lvalue/rvalue.